### PR TITLE
Convert apigateway to sam

### PR DESCRIPTION
**Pull request by Andrewdt97**

### CODE REVIEW

File: .aws-sam/build.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# This file is auto generated by SAM CLI build command
```
Comment (Andrewdt97) [inline .aws-sam/build.toml]: Remove all .aws-sam/ files

File: .gitignore [hunk 2]
```diff
@@ -42,4 +42,6 @@ out
 .DS_Store
 
 */env.json
-*/event.json
\ No newline at end of file
+*/event.json
+
+.aws-sam
```
Comment (Andrewdt97) [inline .gitignore]: Add a `/`

File: backend-lambda/index.js [hunk 3]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
 export const handler = async (event, context) => {
+  const body =
+    typeof event.body === "string" ? JSON.parse(event.body) : event.body;
   console.info("event", event);
   console.info("context", context);
   try {
     if (!pool) {
-      const connectionString = process.env.DATABASE_URL;
+      const connectionString = process.env.DATABASEURL;
+      console.log(connectionString);
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: Remove me

File: backend-lambda/index.js [hunk 4]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
       pool = new pg.Pool({
         connectionString,
-        application_name: "$ docs_lambda_node",
-        max: 1,
+        // application_name: "$ docs_lambda_node",
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: Remove me

File: backend-lambda/index.js [hunk 5]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
+        max: 3,
       });
     }
   } catch (e) {
     console.error(JSON.stringify(e));
   }
   let response;
-
+  // get all products
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: This probably needs to go between 22/23

File: backend-lambda/index.js [hunk 6]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
   if (event.path === "/products") {
     if (event.httpMethod === "GET") {
       response = await productService.getAllProducts(pool);
     } else if (event.httpMethod === "POST") {
-      response = await productService.createProduct(
-        pool,
-        JSON.parse(event.body)
-      );
+      let product = event.body;
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: 1) This be `const` not `let`. 
2) Line 12 parsed `event.body` already, we we want to use that parse and not `event.body` itself.

File: backend-lambda/index.js [hunk 7]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
+      response = await productService.createProduct(pool, product);
     }
-    // create product
-  } else if (event.path.includes("/product/")) {
-    // update product
+    // create new product
+  } else if (event.path === "/product") {
+    response = await productService.createProduct(pool, body);
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: This should also be `updateProduct` and we should also check in line 39 that the method is `PUT`.

File: backend-lambda/index.js [hunk 8]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
+    // update a product as sold
   } else if (event.path === "/products/sold") {
-    // mark products sold
-  } else if (event.path === "/products/to-floor") {
+    response = await productService.updateProduct(pool, body.product_id, body);
```
Comment (Andrewdt97) [inline backend-lambda/index.js:42]: We need to check if this is `POST`

File: backend-lambda/index.js [hunk 9]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
     // make items on floor
+  } else if (event.path === "/products/to-floor") {
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: Check for `POST`

File: backend-lambda/index.js [hunk 10]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
-        pool,
-        JSON.parse(event.body)
-      );
+      let product = event.body;
+      response = await productService.createProduct(pool, product);
     }
-    // create product
-  } else if (event.path.includes("/product/")) {
-    // update product
+    // create new product
```
Comment (Andrewdt97) [inline backend-lambda/index.js:38]: Move this up

File: backend-lambda/index.js [hunk 11]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
+  } else if (event.path === "/product") {
+    response = await productService.createProduct(pool, body);
+    // update a product as sold
```
Comment (Andrewdt97) [inline backend-lambda/index.js:41]: Move this up

File: backend-lambda/index.js [hunk 12]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
   } else if (event.path === "/products/sold") {
-    // mark products sold
-  } else if (event.path === "/products/to-floor") {
+    response = await productService.updateProduct(pool, body.product_id, body);
     // make items on floor
+  } else if (event.path === "/products/to-floor") {
+    response = await productService.updateProduct(pool, body.product_id, body);
+    // get all users
```
Comment (Andrewdt97) [inline backend-lambda/index.js:50]: Move down one line

File: backend-lambda/index.js [hunk 13]
```diff
@@ -9,38 +9,42 @@ const headers = {
...
 
 export const handler = async (event, context) => {
+  const body =
+    typeof event.body === "string" ? JSON.parse(event.body) : event.body;
   console.info("event", event);
   console.info("context", context);
   try {
     if (!pool) {
-      const connectionString = process.env.DATABASE_URL;
+      const connectionString = process.env.DATABASEURL;
```
Comment (Andrewdt97) [inline backend-lambda/index.js]: Add underscore back

File: template.yml [hunk 14]
```diff
@@ -40,62 +49,65 @@ Resources:
       CodeUri: ./backend-lambda
       Environment:
         Variables:
-          DATABASE_URL: !Ref DATABASE_URL
+          DATABASE_URL: !Ref DatabaseUrl
       Events:
```
Comment (Andrewdt97) [inline template.yml]: Do we still need this block if we have the RestApi block?

File: template.yml [hunk 15]
```diff
@@ -40,62 +49,65 @@ Resources:
...
+            Path: /
+            Method: ANY
 
   RestApi:
     Type: AWS::Serverless::Api
     Properties:
       StageName: prod
       Auth:
-        DefaultAuthorizer: IPAuthorizerLambdaFunction
+        DefaultAuthorizer: IpAuthorizerLambdaFunction
```
Comment (Andrewdt97) [inline template.yml]: I think this needs a `!Ref `

File: template.yml [hunk 16]
```diff
@@ -40,62 +49,65 @@ Resources:
...
+          /products:
+            get: &getProducts
               responses:
                 "200":
-                  description: "Successful response for GET request"
+                  description: "Successful response for GET products request"
               x-amazon-apigateway-integration:
-                uri: !Sub "arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${BackendLambdaFunctionArn}/invocations"
+                uri: !Sub "arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${BackendLambda.Arn}/invocations"
                 httpMethod: "GET"
```
Comment (Andrewdt97) [inline template.yml]: This needs to be `POST`

File: template.yml [hunk 17]
```diff
@@ -40,62 +49,65 @@ Resources:
...
+        DefaultAuthorizer: IpAuthorizerLambdaFunction
       DefinitionBody:
         swagger: "2.0"
         info:
           title: "RestApi"
         paths:
-          /myendpoint:
-            get:
+          /products:
+            get: &getProducts
```
Comment (Andrewdt97) [inline template.yml]: Rename to `&standardImplementation` and reuse.

File: template.yml [hunk 18]
```diff
@@ -40,62 +49,65 @@ Resources:
...
-            description: Default response for 5xx errors
-        x-amazon-apigateway-gateway-responses:
-          DEFAULT_4XX:
-            responseTemplates:
-              application/json: '{"message": "Client Error: {statusCode}"}'
-          DEFAULT_5XX:
-            responseTemplates:
-              application/json: '{"message": "Server Error: {statusCode}"}'
+          /product:
+            post: *postProducts
```
Comment (Andrewdt97) [inline template.yml]: Update these methods as commented on in the backend lambda index.js.

File: .aws-sam/build.toml (hunk 1 above)
Comment (ritchieritch42) [reply .aws-sam/build.toml]: done

File: .gitignore (hunk 2 above)
Comment (ritchieritch42) [reply .gitignore]: done

File: backend-lambda/index.js (hunk 13 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 3 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 4 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 5 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: Each comment corresponds to the event.path below it correct?

File: backend-lambda/index.js (hunk 6 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 10 above)
Comment (ritchieritch42) [reply backend-lambda/index.js:38]: Same thing with this one as the mention of moving  // get all products to lines 22-23 - do you want comments for the event.paths to be above each or below?

File: backend-lambda/index.js (hunk 11 above)
Comment (ritchieritch42) [reply backend-lambda/index.js:41]: do you want comments for the event.paths to be above each or below?

File: backend-lambda/index.js (hunk 7 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 8 above)
Comment (ritchieritch42) [reply backend-lambda/index.js:42]: done

File: backend-lambda/index.js (hunk 9 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done

File: backend-lambda/index.js (hunk 12 above)
Comment (ritchieritch42) [reply backend-lambda/index.js:50]: waiting review on comments

File: template.yml (hunk 14 above)
Comment (ritchieritch42) [reply template.yml]: We do not need it, removing.

File: template.yml (hunk 15 above)
Comment (ritchieritch42) [reply template.yml]: It does, adding.

File: template.yml (hunk 17 above)
Comment (ritchieritch42) [reply template.yml]: done

File: template.yml (hunk 16 above)
Comment (ritchieritch42) [reply template.yml]: done

File: backend-lambda/index.js (hunk 5 above)
Comment (Andrewdt97) [reply backend-lambda/index.js]: Yes, but it should be above the last if block before the call. Not sure why I put 22/23, it should be between 29/30.

File: backend-lambda/index.js (hunk 10 above)
Comment (Andrewdt97) [reply backend-lambda/index.js:38]: Above. So this one should be  either between 31/32 (and indenting in-line with the else if clause) or just on the end of 32. Either works.

File: backend-lambda/index.js (hunk 9 above)
Comment (Andrewdt97) [reply backend-lambda/index.js]: There's no reason to nest if statements if ALL the outer block contains is the if block inside. You can do `else if (event.path === '/products/to-floor' && event.httpMethod === 'POST')...`

Comment (Andrewdt97) [reply backend-lambda/index.js]: Fix the other nested ifs too if they apply

File: template.yml (hunk 18 above)
Comment (ritchieritch42) [reply template.yml]: done

File: backend-lambda/index.js (hunk 9 above)
Comment (ritchieritch42) [reply backend-lambda/index.js]: done, recommitted for your review, will leave this unresolved until you approve

File: backend-lambda/index.js (hunk 12 above)
Comment (ritchieritch42) [reply backend-lambda/index.js:50]: done

**Pull request merged**